libkbfs/disk_block_cache_wrapped.go

```diff
@@ -0,0 +1,173 @@
+
+	"github.com/keybase/kbfs/kbfsblock"
+	"github.com/keybase/kbfs/kbfscrypto"
+	"github.com/keybase/kbfs/tlf"
+	"github.com/syndtr/goleveldb/leveldb/errors"
+)
+
+const (
+	workingSetCacheFolderName = "kbfs_block_cache"
+	syncCacheFolderName       = "kbfs_sync_cache"
+)
+
+// diskBlockCacheConfig specifies the interfaces that a DiskBlockCacheStandard
+// needs to perform its functions. This adheres to the standard libkbfs Config
+// API.
+type diskBlockCacheConfig interface {
+	codecGetter
+	logMaker
+	clockGetter
+	diskLimiterGetter
+	syncedTlfGetterSetter
+}
+
+type diskBlockCacheWrapped struct {
+	config          diskBlockCacheConfig
+	workingSetCache DiskBlockCache
+	syncCache       DiskBlockCache
+}
+
+var _ DiskBlockCache = (*diskBlockCacheWrapped)(nil)
+
+func newDiskBlockCacheWrapped(config diskBlockCacheConfig, storageRoot string) (
+	cache *diskBlockCacheWrapped, err error) {
+	workingSetCacheRoot := filepath.Join(storageRoot, workingSetCacheFolderName)
+	workingSetCache, err := newDiskBlockCacheStandard(config, false,
+		workingSetCacheRoot)
+	if err != nil {
+		return nil, err
+	}
+	return &diskBlockCacheWrapped{
+		config:          config,
+		workingSetCache: workingSetCache,
+		syncCache:       nil,
+	}, nil
+}
+
+// Get implements the DiskBlockCache interface for diskBlockCacheWrapped.
+func (cache *diskBlockCacheWrapped) Get(ctx context.Context, tlfID tlf.ID,
+	blockID kbfsblock.ID) (
+	buf []byte, serverHalf kbfscrypto.BlockCryptKeyServerHalf,
+	hasPrefetched bool, err error) {
+	// TODO: add mutex to guard sync state
+	primaryCache := cache.workingSetCache
+	secondaryCache := cache.syncCache
+	if cache.config.IsSyncedTlf(tlfID) {
+		primaryCache = cache.syncCache
+		secondaryCache = cache.workingSetCache
+	}
+	// Check both caches if the primary cache doesn't have the block.
+	buf, serverHalf, hasPrefetched, err = primaryCache.Get(ctx, tlfID, blockID)
+	if _, isNoSuchBlockError := err.(NoSuchBlockError); isNoSuchBlockError {
+		if secondaryCache != nil {
+			return secondaryCache.Get(ctx, tlfID, blockID)
+		}
+	}
+	return buf, serverHalf, hasPrefetched, err
+}
+
+// GetMetadata implements the DiskBlockCache interface for
+// diskBlockCacheWrapped.
+func (cache *diskBlockCacheWrapped) GetMetadata(ctx context.Context,
+	blockID kbfsblock.ID) (metadata DiskBlockCacheMetadata, err error) {
+	// TODO: add mutex to guard sync state
+	if cache.syncCache != nil {
+		md, err := cache.syncCache.GetMetadata(ctx, blockID)
+		switch err {
+		case nil:
+			return md, nil
+		case errors.ErrNotFound:
+		default:
+			return md, err
+		}
+	}
+	return cache.workingSetCache.GetMetadata(ctx, blockID)
+}
+
+// Put implements the DiskBlockCache interface for diskBlockCacheWrapped.
+func (cache *diskBlockCacheWrapped) Put(ctx context.Context, tlfID tlf.ID,
+	blockID kbfsblock.ID, buf []byte,
+	serverHalf kbfscrypto.BlockCryptKeyServerHalf) error {
+	if cache.config.IsSyncedTlf(tlfID) {
+		cache.workingSetCache.Delete(ctx, []kbfsblock.ID{blockID})
```

Comment:
I was torn about this: on the one hand I wanted it to be atomic to minimize disk duplication, but I think you're right, especially since LevelDB doesn't guarantee immediate disk cleanup. I'll put the `Delete`s in a goroutine.